### ISSUE TITLE
Incorrect warning and `type annotations needed` error for types defined using raw identifiers

### ISSUE THREAD

**Issue by olegnn** [rust-lang/rust#69052]
This code

```rust
fn r<r#fn>() {}

fn main() {
    r();
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a12ec17055cf3c08f4b6d59b31969dfc))

Gives incorrect warning and error for type declared using raw identifiers:

```
warning: type parameter `fn` should have an upper camel case name
 --> src/main.rs:1:6
  |
1 | fn r<r#fn>() {}
  |      ^^^^ help: convert the identifier to upper camel case: `Fn`
  |
  = note: `#[warn(non_camel_case_types)]` on by default

error[E0282]: type annotations needed
 --> src/main.rs:4:5
  |
4 |     r();
  |     ^ cannot infer type for type parameter `fn`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground`.
```

**Comment by jonas-schievink** [comment]
These seem correct to me, since `r#` is just an escape for keywords, not part of the actual name. Otherwise `#[no_mangle]` should also create a symbol called `r#something`.

**Comment by Centril** [comment]
The type checking error here is correct. You have a type parameter that is passed no argument so it gets an inference variable that is never solved.

As for the warning, can you elaborate on why it is incorrect? It seems like it is responding to the fact that the keyword is lower case, which is correct behavior. If you use `fn r<r#Fn>() {}` the warning goes away.

**Comment by olegnn** [comment]
@Centril issue isn't about incorrect type checking but about incorrect messages
```
type parameter `fn` should have an upper camel case name
```
and
```
cannot infer type for type parameter `fn`
```

**Comment by olegnn** [comment]
should be `r#fn` both

**Comment by Centril** [comment]
Oh I see; now I get what you're saying; it's a diagnostics issue, not a behavior issue. Yes, that seems right.

**Comment by olegnn** [comment]
I will change type parameter names to make them look more obvious.

**Comment by estebank** [comment]
Current output:

```
error[[E0282]](https://doc.rust-lang.org/nightly/error_codes/E0282.html): type annotations needed
 --> src/main.rs:4:5
  |
4 |     r();
  |     ^ cannot infer type of the type parameter `fn` declared on the function `r`
  |
help: consider specifying the generic argument
  |
4 |     r::<fn>();
  |      ++++++
```

We'll need another repro case for a message not including `r#` in the token name (I think there are other tickets about it).
